### PR TITLE
Config refactor

### PR DESCRIPTION
**Pull request by t-keazirian**
Created a Routes class:
1. Creates a _routes dictionary that holds the path and the whole Route object
2. Methods to add a route, get the route, check if the path exists in the dictionary, and handle the response for that route

Created Route class:
1. Holds the Path, List of Methods allowed for that path, and the Handler
2. Method to check if method exists for that path

Refactored Router class:
1. Router class now takes the Routes object (not routes config)

Updated tests:
1. Tests use only MockHandler

Program:
1. Moved adding/constructing routes to Program to pass into Server (where RoutesConfig was)

** Removed all AllowedHttpMethods from each Handler **

### CODE REVIEW

File: TKeazirian.HTTPServer/Response/HttpMethods.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+namespace TKeazirian.HTTPServer.Response;
+
+public static class HttpMethods
+{
+    public static readonly string Get = "GET";
+    public static readonly string Post = "POST";
+    public static readonly string Options = "OPTIONS";
+    public static readonly string Head = "HEAD";
+    public static readonly string Put = "PUT";
+    public static readonly string Patch = "PATCH";
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Response/HttpMethods.cs:10]: this is a begining of a great idea. I see `string` and `List<string>` all throughout the code. Is it time to introduce a Type?  This way, instead of passing strings around, you can use just these methods?

File: TKeazirian.HTTPServer/Router/Route.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+namespace TKeazirian.HTTPServer.Router;
+
+using Handler;
+
+public class Route
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Route.cs:5]: A Route as an object. I wonder if this is useful, or adding a level on unneeded complexity.

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 3]
```diff
@@ -6,80 +6,56 @@ namespace TKeazirian.HTTPServer.Router;
...
 
-    public Router(RoutesConfig routesConfig)
+    public Router(Routes routes)
     {
-        _routesConfig = routesConfig;
+        _routes = routes;
     }
 
-    public Response Route(Request request)
+    public Response GetResponse(Request request)
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Router.cs:16]: GetResponse, getResponse ... a GET Response.  These are three different things. It can be confusing to keep track of.  I like that name `Route` better in this case. What does a Router do? it Routes, using a method called Route.

File: TKeazirian.HTTPServer/Router/Route.cs (hunk 2 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Route.cs:5]: To me, it seemed like the best way to hold the data and allows for me to have the path, methods, and handler all in one object.

File: TKeazirian.HTTPServer/Program.cs [hunk 4]
```diff
@@ -1,14 +1,40 @@
...
+            new Route("/redirect", new List<string>() { HttpMethods.Get }, new RedirectHandler()));
+        routes.AddRoute(
+            new Route("/head_request", new List<string>() { HttpMethods.Get }, new SimpleHeadHandler()));
+        routes.AddRoute(
+            new Route("/method_options", new List<string>() { HttpMethods.Get }, new SimpleOptionsHandler()));
+        routes.AddRoute(
+            new Route("/method_options2", new List<string>() { HttpMethods.Get, HttpMethods.Put, HttpMethods.Post },
+                new SimpleOptionsHandler()));
+        return routes;
+    }
```
[the remark this answers is not in the document]
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Program.cs:39]: Went a little too far with removing it - I will re-create it and put this in there!

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs [hunk 5]
```diff
@@ -1,30 +1,33 @@
...
+        routes.AddRoute(
+            new("/redirect", new() { Get }, new RedirectHandler()));
+        routes.AddRoute(
+            new("/head_request", new() { Get }, new SimpleHeadHandler()));
+        routes.AddRoute(
+            new("/method_options", new() { Get }, new SimpleOptionsHandler()));
+        routes.AddRoute(
+            new("/method_options2", new() { Get, Put, Post },
+                new SimpleOptionsHandler()));
+        return routes;
```
Comment (tpetersen0308) [inline TKeazirian.HTTPServer/Router/RoutesConfig.cs:31]: This looks much more like what I'd expect to see in a routes configuration! It's kind of similar to Express, where you'd defined it something like `router.get("/simple_get", simpleGetHandler)`

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs [hunk 6]
```diff
@@ -1,30 +1,33 @@
...
-    {
-        Routes = routes;
+        routes.AddRoute(
+            new("/simple_get", new() { Get }, new SimpleGetHandler())
+        );
+        routes.AddRoute(
+            new("/simple_get_with_body", new() { Get }, new SimpleGetHandler())
+        );
+        routes.AddRoute(
+            new("/echo_body", new() { Post }, new EchoBodyHandler()));
```
Comment (tpetersen0308) [inline TKeazirian.HTTPServer/Router/RoutesConfig.cs]: since you have the path configured here in the dictionary, do you need it in the `Route` instance as well? Do you actually reference `Route.path`?

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/RoutesConfig.cs]: This was changed to be more explicit.  The IDE had suggested removing the types.

Now it looks like this: 

```csharp
new Route("/echo_body", new List<HttpMethod> { HttpMethod.POST }, new EchoBodyHandler()));
```

File: TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs [hunk 7]
```diff
@@ -1,3 +1,5 @@
+using TKeazirian.HTTPServer.Response;
+
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs:2]: Is this being used?

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs [hunk 8]
```diff
@@ -1,30 +1,32 @@
...
+            new Route("/echo_body", new List<HttpMethod> { HttpMethod.POST }, new EchoBodyHandler()));
+        routes.AddRoute(
+            new Route("/redirect", new List<HttpMethod> { HttpMethod.GET }, new RedirectHandler()));
+        routes.AddRoute(
+            new Route("/head_request", new List<HttpMethod> { HttpMethod.GET }, new SimpleHeadHandler()));
+        routes.AddRoute(
+            new Route("/method_options", new List<HttpMethod> { HttpMethod.GET }, new SimpleOptionsHandler()));
+        routes.AddRoute(
+            new Route("/method_options2", new List<HttpMethod> { HttpMethod.GET, HttpMethod.PUT, HttpMethod.POST },
+                new SimpleOptionsHandler()));
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/RoutesConfig.cs:29]: I notice there is not a 1:1 relationship between a path, and an handler. Instead you have two paths pointing to the same handler. This works, but for a large project with multiple endpoints, it could really come back to bite you.  Same goes the SimpleGetHandler reuse above.  

I know this is really a fault of how the acceptance test suite is coded, but I wanted to mention it here just so you are aware.

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs (hunk 6 above)
Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Router/RoutesConfig.cs]: That def reads better! I think what I was getting at though is the the path ends up getting set in two places: as the key in the router dictionary, and also as a property on the `Route` itself. I think it's not particularly problematic, since it seems to only be set in one place. Otherwise my concern would be that you could end up with some discrepancy between the path in the key and the path configured on the route the key points to. But that doesn't seem to be a real issue here, so my real question is, is the path actually doing anything important as a property on the route instance, since it's already represented in the router?

TL;DR: are there actually any dependencies on the path property of the route class? Or is it essentially cruft?

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/RoutesConfig.cs]: I reference `route.Path` in Routes.cs (line 17).

```csharp
_routes.Add(route.Path, route);
```   

That sets the whole Route object, adding a Route to Routes.  As you mentioned, I'm only ever setting the path in the `AddRoute()` method in the RoutesConfig.  But I'm not sure, maybe I don't understand fully what you're saying?

File: TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs (hunk 7 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs]: 🔫 bye bye

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs (hunk 8 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/RoutesConfig.cs:29]: Got it.  So normally (for a large project), the `/method_options` (for example) would have one handler, and then the `/method_options2` would have a separate handler?  One handler per one path, correct?  

There was [feedback](https://github.com/t-keazirian/HTTPServer/pull/27#discussion_r914256806) on by last PR where Travis was asking why I had two handlers - maybe he wasn't getting at only having one handler, but that was my impression.

File: TKeazirian.HTTPServer/Router/RoutesConfig.cs (hunk 6 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/RoutesConfig.cs]: Okay okay, I got it now - I understand!  Will update PR soon.

File: TKeazirian.HTTPServer/Router/Routes.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+using Request;
+
+public class Routes
+{
+    private readonly Dictionary<string, Route> _routes;
+
+    public Routes()
+    {
+        _routes = new Dictionary<string, Route>();
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Routes.cs:13]: Is `Routes` just a `List<Route>` in disguise?  Could you accomplish everything here by just using a List?

Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Routes.cs:13]: That is, assuming a `Route` includes a `path`, which it appears it no longer does.

Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Routes.cs:13]: IF you added it back, could you just use a `List<Route>` and drop this class though?

Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Routes.cs:13]: that was my original question; not why the `path` was removed.

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Routes.cs:13]: Hmm...tbh I'm not sure but I will look at it to see!

**Review by tpetersen0308 — APPROVED** [review]

File: TKeazirian.HTTPServer/Router/Routes.cs (hunk 9 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Routes.cs:13]: So I played around with this - got it working, unit tests passing but acceptance tests not passing.

I may be approaching it in an unexpected way, but it feels more complicated to do it this way?  I think I would have to loop through the whole List<Route> and then check each one according to the logic , as opposed to the way now which feels more direct.  Especially if I remove the path from the Route object.
